### PR TITLE
[OpenCL] Fix issues with brackets for if-conditions in for-loops

### PR DESCRIPTION
**Pull request by mairooni**
#### Description

This PR fixes the closing brackets for OpenCL kernels that contain if-conditions either before or in for-loops. 

#### Problem description

It was observed that in some cases where there was an if-condition in a kernel that contained for-loop(s)  (e.g. before a for-loop, in an inner for-loop etc.) , either some closing brackets were missing or unnecessary additional closing brackets were included. With this PR, the code generation examines these cases and generates the correct brackets.

#### Backend/s tested

Mark the backends affected by this PR.

- [x] OpenCL
- [ ] PTX
- [ ] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [x] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [x] No

#### How to test the new patch?
Run `tornado-test -V -pk uk.ac.manchester.tornado.unittests.branching.TestLoopConditions`
All unittests have been tested along with RayTracer and KFusion.

### CODE REVIEW

**Comment by stratika** [comment]
Just a note, I tested the branch in `macOSx` and `Linux Ubuntu 23.10` and the fix is working.

**Review by stratika — APPROVED** [review]

**Comment by mairooni** [comment]
> Just a note, I tested the branch in `macOSx` and `Linux Ubuntu 23.10` and the fix is working.

Thanks for testing it Thano!

**Comment by jjfumero** [comment]
That's great. Based on my experience when we fix the OpenCL visitor is that we always need to check with KFusion.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/branching/TestLoopConditions.java [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+ * <code>
+ * tornado-test -V uk.ac.manchester.tornado.unittests.branching.TestLoopConditions
+ * </code>
+ */
+public class TestLoopConditions extends TornadoTestBase {
+
+    public static void conditionBeforeSingleLoopReturn(IntArray a, IntArray b, IntArray c) {
+        if (a.get(0) > b.get(0)) {
+            return;
+        }
```
Comment (jjfumero) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/branching/TestLoopConditions.java:48]: What is the expected behaviour of this type of code when running in parallel on a GPU?  We can't stop the GPU kernel.

File: tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/branching/TestLoopConditions.java [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+        for (@Parallel int i = 0; i < c.getSize(); i++) {
+            c.set(i, a.get(i) + b.get(i));
+        }
+    }
+
+    public static void conditionInSingleLoopReturn(IntArray a, IntArray b, IntArray c) {
+        for (@Parallel int i = 0; i < c.getSize(); i++) {
+            if (a.get(0) > b.get(0)) {
+                return;
+            }
```
Comment (jjfumero) [inline tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/branching/TestLoopConditions.java:67]: Same here, what is the expected behaviour when running on a GPU?

**Comment by stratika** [comment]
> That's great. Based on my experience when we fix the OpenCL visitor is that we always need to check with KFusion.

Good note! Just tested `kfusion` with OpenCL, PTX on NVIDIA RTX A2000 GPU.

I also tested `TornadoVM-Ray-Tracer` for OpenCL.

**Review by jjfumero — APPROVED** [review]
After the sync, let's clarify in the header of the test and before each kernel why this test is valid and under which conditions will be still valid for TornadoVM.

**Review by jjfumero — APPROVED** [review]

**Pull request merged**